### ISSUE TITLE
Offline Map Update: Old maps not properly removed until OsmAnd is closed

### ISSUE THREAD

**Issue by BenBE** [osmandapp/Osmand#6590]
I noticed the following issue while using OsmAnd+ (nightly 2019-02-22, but not specific to that):

> What version of OsmAnd are you using, on what device and Android/iOS version?
Last checked with OsmAnd+ nightly from 2019-02-22, but noticed since quite a while.

OsmAnd Version: OsmAnd~ 3.3.0#26819m 2019-02-22 (from Nightly website)
Android Version: 4.3.1
Phone: Samsung Galaxy S4 (jfltexx)

> Do you use the offline maps offered within the OsmAnd app for download, or online (tile / raster) maps?
Offline maps.

> If you have an issue related to offline maps, tell us the exact name of the map file where the issue occurs, and its edition date.
Independent of the actual map file.

> Exact issue description:
When /updating/ existing offline maps there seems to be an issue with releasing all open references to the old map file handles. This causes the disk space of the old map file to be kept blocked event after the new map file version has been successfully downloaded/installed.

> Steps to reproduce:
0. Note down the free memory on the storage device associated with the OsmAnd offline maps folder.
1. Update any installed offline map file (best noticed with large maps like the Netherlands standard map or the large state maps from Germany like Bavaria)
2. Compare the free memory on the storage device associacted with the OsmAnd offline maps folder
3. Notice that approximately the size of the old map file is used more than expected.
4. Close OsmAnd with the task manager
5. Compare free memory again and notice memory for the old offline maps has finally been freed.

> Workaround:
Force-close OsmAnd in the task manager after the map update has completed and wait a few seconds.

> References:
This is no duplicate of #5541, despite it looking similar: Looking into the affected folders while OsmAnd is still running I can see several files still sitting there.

**Comment by aceman444** [comment]
Yes, this existed for a long time, and I think it was already reported somewhere. Maybe on the old issue tracker before github.

**Comment by aceman444** [comment]
This is also clearly visible in OsmAnd live. If you click a country to update, the dialog shows "Size of update", which is the total size of files already downloaded in the Osmand Live folder, for that country. If you click "update now", the size will increase radically. Then when you kill Osmand and revisit the dialog, the size reduces to the real size of files in the osmand/data/files/live folder.

**Comment by vshcherb** [comment]
Yes it is old issue though it was different for different devices / OS. After couple of investigation rounds we noticed that the leak is not in the OsmAnd code though we didn't find solution how to force or workaround it.

**Comment by Avispa** [comment]
Let me say "_I can confirm that_" to push this at least 2 years old issue a little. (Even though I think as well as mentioned by **aceman444**, this issue exist much much longer.)

Android Version: 9 (maybe 9.0? No minor displayed.)
OSMAnd Version: 3.9.10 (F-Droid)

This is quite cumbersome on devices with small storage space since you have to update every map singularly: update -- quit -- kill -- restart -- loop over again.

In the worst case scenario it means you cannot update a map because it needs twice the space than the file itself would occupy,. Which is for Germany/Bavaria for example quite a killer.

**Comment by cboltz** [comment]
I can confirm this problem with the latest version on Android (via F-Droid).

> Yes it is old issue though it was different for different devices / OS. After couple of investigation rounds we noticed that the leak is not in the OsmAnd code though we didn't find solution how to force or workaround it.

Wild guess - could it be that the old map files are still open()ed somehow? As long as this is true, they can be deleted (as in "remove the entry from the directory") - but their disk space won't be freed until all open file handles for the deleted file get closed.
